File: common/src/main/scala/se/kth/id2203/link/NetworkAddress.scala
package se.kth.id2203.link

import com.google.common.collect.ComparisonChain
import com.google.common.primitives.UnsignedBytes
import java.net.{InetAddress, InetSocketAddress}

import se.sics.kompics.network.Address

class NetworkAddress(val isa: InetSocketAddress) extends Address with Comparable[NetworkAddress] {

  def getIp: InetAddress = this.isa.getAddress

  def getPort: Int = this.isa.getPort

  def asSocket: InetSocketAddress = this.isa

  def sameHostAs(other: Address): Boolean = this.asSocket equals other.asSocket

  override final def toString: String = s"${this.isa.getAddress.getHostAddress}:${this.isa.getPort.toString}"

  override def equals(obj: Any): Boolean = obj match {
    case other: NetworkAddress =>
      0 == this.compareTo(other)
    case _ => false
  }

  def compareTo(that: NetworkAddress): Int = ComparisonChain
    .start
    .compare(this.getIp.getAddress, that.getIp.getAddress, UnsignedBytes.lexicographicalComparator)
    .compare(this.getPort, that.getPort)
    .result

}

File: common/src/main/java/se/kth/id2203/link/NetworkAddress.java
package se.kth.id2203.link;

import com.google.common.collect.ComparisonChain;
import com.google.common.primitives.UnsignedBytes;
import java.io.Serializable;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import se.sics.kompics.network.Address;

public class NetworkAddress implements Serializable, Address, Comparable<NetworkAddress> {

    private final static long serialVersionUID = 2536770490757392511L;
    private final InetSocketAddress isa;

    public NetworkAddress(InetAddress addr, int portI) {
        this.isa = new InetSocketAddress(addr, portI);
    }

    @Override
    public InetAddress getIp() {
        return this.isa.getAddress();
    }

    @Override
    public int getPort() {
        return this.isa.getPort();
    }

    @Override
    public InetSocketAddress asSocket() {
        return this.isa;
    }

    @Override
    public boolean sameHostAs(Address other) {
        return this.isa.equals(other.asSocket());
    }

    @Override
    public final String toString() {
        return isa.toString();
    }

    @Override
    public int hashCode() {
        int hash = 3;
        hash = ((11 * hash) + ((this.isa != null) ? this.isa.hashCode() : 0));
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (!(obj instanceof NetworkAddress)) {
            return false;
        }
        final NetworkAddress that = ((NetworkAddress) obj);
        return this.compareTo(that) == 0;
    }

    @Override
    public int compareTo(NetworkAddress that) {
        return ComparisonChain.start()
                .compare(this.isa.getAddress().getAddress(), that.isa.getAddress().getAddress(), UnsignedBytes.lexicographicalComparator())
                .compare(this.isa.getPort(), that.isa.getPort())
                .result();
    }
}